### ISSUE TITLE
Addition of LINE App / Corporation Protocol

### ISSUE THREAD

**Issue by sharonenoch** [ntop/nDPI#1758]
Hello

Is it possible to add support for the LINE app by just importing the following ASN https://ipinfo.io/AS38631 in asn_update.sh and adding the following in the servername : *.line-apps.com and Certificate match CN=*.line-apps.com in ndpi_content_match.c? 

Do we need to write a specific protocol dissector for just the above alone?  

Ideally would like to detect it as a VOIP call, as the UDP gets decoded as RTP..

By adding the above changes in the code I was able to get the below when running ndpiReader. Was not sure if it was complete..

root@kickseed:~/nDPI/example# ./ndpiReader -i /root/line_withlaptopinlansetup.pcap -v2 | grep Line
        **Line                 packets: 37            bytes: 9480          flows: 1**
        3       TCP 10.70.99.12:52604 <-> 147.92.165.206:443 [proto: 91/TLS][**IP: 315/Line**][Encrypted][Confidence: DPI][cat: Web/5][49 pkts/5515 bytes <-> 61 pkts/28606 bytes][Goodput ratio: 52/88][252.54 sec][bytes ratio: -0.677 (Download)][IAT c2s/s2c min/avg/max/stddev: 0/0 6585/4357 59917/60007 16012/13689][Pkt Len c2s/s2c min/avg/max/stddev: 54/60 113/469 681/1514 125/615][Plen Bins: 0,52,3,6,0,1,1,0,4,1,1,0,0,0,0,0,0,0,0,3,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,25,0,0]
        8       TCP 10.70.99.12:55168 <-> 147.92.242.232:443 [proto: 91.315/**TLS.Line**][IP: 315/Line][Encrypted][Confidence: DPI][cat: Chat/9][16 pkts/4057 bytes <-> 21 pkts/5423 bytes][Goodput ratio: 78/78][91.93 sec][Hostname/SNI: uts-front.line-apps.com][bytes ratio: -0.144 (Mixed)][IAT c2s/s2c min/avg/max/stddev: 0/0 7621/4099 30000/30000 12829/10097][Pkt Len c2s/s2c min/avg/max/stddev: 54/60 254/258 627/1514 230/419][Risk: ** TLS (probably) Not Carrying HTTPS **][Risk Score: 10][Risk Info: No ALPN][TLSv1.2][JA3C: ca75ea4a95a9164cc96e372d7d075183][**ServerNames: *.line-apps.com,line-apps.com**][JA3S: 567bb420d39046dbfd1f68b558d86382][Issuer: C=BE, O=GlobalSign nv-sa, CN=GlobalSign RSA OV SSL CA 2018][Subject: C=JP, ST=Tokyo-to, L=Shinjuku-ku, O=LINE Corporation, **CN=*.line-apps.com**][Certificate SHA-1: 3C:37:D7:AB:BE:E6:5A:A5:BE:14:62:C8:21:8C:BC:E3:3E:A8:3E:96][Firefox][Validity: 2020-08-17 06:21:02 - 2022-11-13 12:00:00][Cipher: TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384][Plen Bins: 5,15,5,0,0,15,0,0,5,15,5,0,0,0,0,0,10,10,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,10,0,0]
        23      TCP 147.92.191.86:443 <-> 10.70.99.12:55167 [proto: 91/TLS][**IP: 315/Line**][Encrypted][Confidence: DPI][cat: Web/5][3 pkts/205 bytes <-> 2 pkts/108 bytes][Goodput ratio: 15/0][18.12 sec][Plen Bins: 100,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
        1       UDP 10.70.99.12:56946 <-> 147.92.169.100:17077 [proto: 0/Unknown][**IP: 315/Line**][ClearText][Confidence: Unknown][47819 pkts/43400193 bytes <-> 50866 pkts/43917779 bytes][Goodput ratio: 95/95][221.19 sec][bytes ratio: -0.006 (Mixed)][IAT c2s/s2c min/avg/max/stddev: 0/0 3/3 277/6425 7/32][Pkt Len c2s/s2c min/avg/max/stddev: 62/62 908/863 1098/1098 287/327][PLAIN TEXT (dsslcd)][Plen Bins: 1,13,9,0,1,1,1,0,0,0,0,0,0,0,0,0,13,0,0,0,0,0,0,0,0,0,0,2,0,0,0,47,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
        2       UDP 10.70.99.12:56955 <-> 147.92.169.100:17077 [proto: 0/Unknown][**IP: 315/Line**][ClearText][Confidence: Unknown][29 pkts/5045 bytes <-> 29 pkts/5856 bytes][Goodput ratio: 76/79][17.58 sec][bytes ratio: -0.074 (Mixed)][IAT c2s/s2c min/avg/max/stddev: 0/0 567/567 2042/2042 730/730][Pkt Len c2s/s2c min/avg/max/stddev: 173/200 174/202 201/256 5/10][Plen Bins: 0,0,0,0,98,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]

**Comment by IvanNardi** [comment]
As a general rule, yes, it is enough.
Some thoughts:
* if you want to proper identify the UDP flows (i.e. something like `[proto: 315/Line][IP: 315/Line]`) you need some code somewhere; the exact place depends of the characteristic of these flows: you may need a new dissector or extend an existing one (STUN?).  I don't know this kind of traffic, so I can't be more precise, sorry
* it seems that the line "CN=.line-apps.com" is useless (since the matching is via SNI)

Are you going to push a PR?

**Comment by sharonenoch** [comment]
Thank @IvanNardi .. Yes I can submit a PR.. 

The UDP currently gets decoded as RTP during a VOIP call between 2 LINE apps. Not sure if there is any extra identifier but I will go through it in detail today.. There are no STUN packets that we see.. unlike in teams, zoom etc.. Is there a location where I can upload a pcap, so that you can take a quick look of the VOIP call?.. 

Also I noticed that WHATSAPP and WHATSAPPCALL were different.. the latter gets detected as STUN as expected and the former is probably for CHAT. So do we need to separate the LINE and LINECALL separately?..  If so I need to go back and capture a pcap with just some chat messages going back and forth and look into that also..

**Comment by IvanNardi** [comment]
You can attached the pcap directly here (as zip or tgz file)

**Comment by IvanNardi** [comment]
> So do we need to separate the LINE and LINECALL separately?

If you are able to tell LINE from LINECALL, it will be useful. Otherwise only LINE is fine

**Comment by sharonenoch** [comment]
[Line_withoutdelay.zip](https://github.com/ntop/nDPI/files/9677863/Line_withoutdelay.zip)  

@IvanNardi 
Attached the zip file of the LINE call....   This filter should give you all the packets.. ip.dst == 147.92.128.0/17 || ip.src == 147.92.128.0/17  Also in Wireshark, can enable decode UDP as RTP

I will try to submit the PR today for the LINE, if not tomorrow.

**Comment by sharonenoch** [comment]
I was able to make progress as you suggested by extending the existing protocol of RTP by adding the following code inside the section where they check the RTP flow in rtp.c..  /* Check whether this is an RTP flow */

But the issue is there are a bunch of 40 or more UDP packets at the beginning between the 2 IPs, so I believe the code is reaching some depth of packets after which it does not check the flow anymore to determine the interior protocol?..

if(flow->guessed_protocol_id_by_ip == NDPI_PROTOCOL_LINE){
        ndpi_set_detected_protocol(ndpi_struct, flow, NDPI_PROTOCOL_LINE, NDPI_PROTOCOL_UNKNOWN, NDPI_CONFIDENCE_DPI);
        return;
      }

So to ensure the above code would work, I stripped the pcap to just have the RTP packets alone and then it worked fine and the nDPIreader printed the below

Detected protocols:
        Line                 packets: 23738         bytes: 20217673      flows: 2


1       UDP 10.200.3.125:51161 <-> 147.92.169.90:29070 **[proto: 315/Line][IP: 315/Line]**[Encrypted][Confidence: DPI][cat: Chat/9][11041 pkts/9875746 bytes <-> 12626 pkts/10279638 bytes][Goodput ratio: 95/95][62.52 sec][bytes ratio: -0.020 (Mixed)][IAT c2s/s2c min/avg/max/stddev: 0/0 3/3 98/199 8/7][Pkt Len c2s/s2c min/avg/max/stddev: 72/72 894/814 1098/1096 279/363][Plen Bins: 0,18,9,29,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,16,24,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
        2       UDP 147.92.169.90:23979 -> 10.200.3.59:64862 **[proto: 315/Line][IP: 315/Line]**[Encrypted][Confidence: DPI][cat: Chat/9][71 pkts/62289 bytes -> 0 pkts/0 bytes][Goodput ratio: 95/0][57.99 sec][bytes ratio: 1.000 (Upload)][IAT c2s/s2c min/avg/max/stddev: 0/0 698/0 5992/0 1291/0][Pkt Len c2s/s2c min/avg/max/stddev: 108/0 877/0 1084/0 287/0][Risk: ** Unidirectional Traffic **][Risk Score: 10][Risk Info: No client to server traffic][PLAIN TEXT (.rE.Z@)][Plen Bins: 0,0,5,7,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,2,1,0,8,16,27,14,11,5,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]

Most of the line call dumps have that same behavior where the RTP starts only after about 40 or so UDP packets exchange.. The packet length though seem to be around 540, 541 and 542.

**Comment by IvanNardi** [comment]
There is an upper limit on the number of pkts processed per flow.
Detecting voip calls doesn't seem trivial. If you prefer/want you can push the "standard" part now and we will see next what we are able to do about the voip part

**Comment by sharonenoch** [comment]
@IvanNardi  Ok I created a pull request for the standard part.. 

#1759 

I am not sure why the h323_overflow.pcap.out is differing with 316 vs 317.. When I do a make check locally on my branch in my ubuntu machine it generates as 316 and everything passes. I thought it was the extra lines and then removed those extra lines.. It succeeded for one do.sh check and failed in another...  Any pointers here would be great..


Is the upper limit 25? I added some prints for packet_counter and saw that it was exiting after 25.. I was trying to see where it was defined or checked in the code but could not find the exact location where it is checked..

Also I may have found a pattern which may not have to depend on RTP decode for the VOIP at least as an initial check..

For LINE Chat -> Its always TCP and TLS.. no UDP

For LINE VOIP Call -> Its always UDP and RTP as we are already aware.. But multiple LINE pcaps show an interesting payload len in the reply packet from their servers to the clients for the initial set of UDP packets which are not RTP. After the first packet, the len in this direction is always 542 and the number of packets in this sequence are 16. I am not sure if this is a coincidence since we had 4 separate line call captures from 2 different machines via entirely different routers. 

So was wondering if we can have a initial dissector like below and the extension of the RTP dissector just in case they reduce the number of UDP packets in scenarios.

if (udp){
    if(guessed protocol by ip == LINE ){
        if packet_counter > 2 && packet_direction_counter > 2  
          /* if there is a way to check a sequence of packets with len 542 by storing a private context somewhere it would be better 
             rather than a single check for len 542 */
            if payload_packet_len == 542
    }
}

**Comment by IvanNardi** [comment]
It seems that the patterns are more complex. As attachment you can find an example of Line call I have been able to do: the general pattern is the same (some "random" UDP before RTP) but details are quite different
[line_voip.zip](https://github.com/ntop/nDPI/files/9690027/line_voip.zip)

**Comment by sharonenoch** [comment]
Yes looks like.. In your dump it was routing it via Naver Corp as the 125.209.x.x which is not part of the LINE Corp ASN..  Looks like Line is owned by both Soft Bank and Naver Corp..

Was it an audio call or video call.. All the dumps which I had checked were video calls.. Will try an audio call also and see the difference.

**Comment by IvanNardi** [comment]
@sharonenoch , could you test if #1761 works for you, please?

**Comment by sharonenoch** [comment]
@IvanNardi  Yes it works.. Checked against the 3 Video pcaps that I had.. it detected the Line Call..  Good eye on the 4th byte incrementing  sequence..  

```
root@kickseed:~/sharonnDPI/nDPI# ./example/ndpiReader -i /root/linepcap/line_ndpi.pcap -v2 | grep LineCall
        
LineCall             packets: 24415         bytes: 20278902      flows: 2

        1       UDP 10.200.3.125:51161 <-> 147.92.169.90:29070 [proto: 316/LineCall][IP: 315/Line][Encrypted][Confidence: DPI][cat: VoIP/10][11410 pkts/9936925 bytes <-> 12995 pkts/10340033 bytes][Goodput ratio: 95/95][70.07 sec][bytes ratio: -0.020 (Mixed)][IAT c2s/s2c min/avg/max/stddev: 0/0 3/4 225/7269 8/71][Pkt Len c2s/s2c min/avg/max/stddev: 62/62 871/796 1098/1096 304/374][Plen Bins: 1,17,15,23,1,1,0,0,0,0,0,0,0,0,0,0,13,0,0,0,0,0,0,0,0,0,0,0,0,0,5,21,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
        4       UDP 10.200.3.125:51170 <-> 147.92.169.90:29070 [proto: 316/LineCall][IP: 315/Line][Encrypted][Confidence: DPI][cat: VoIP/10][5 pkts/898 bytes <-> 5 pkts/1046 bytes][Goodput ratio: 77/80][8.07 sec][bytes ratio: -0.076 (Mixed)][IAT c2s/s2c min/avg/max/stddev: 1999/1999 2009/2009 2038/2037 17/16][Pkt Len c2s/s2c min/avg/max/stddev: 174/198 180/209 202/254 11/22][Plen Bins: 0,0,0,0,80,10,10,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]

root@kickseed:~/sharonnDPI/nDPI# ./example/ndpiReader -i /root/linepcap/line_withdelay.pcap -v2 | grep LineCall
        
LineCall             packets: 103908        bytes: 91037590      flows: 5

        1       UDP 10.70.99.11:53949 <-> 147.92.169.111:10583 [proto: 316/LineCall][IP: 315/Line][Encrypted][Confidence: DPI][cat: VoIP/10][17061 pkts/11239749 bytes <-> 58279 pkts/55850608 bytes][Goodput ratio: 94/96][309.20 sec][bytes ratio: -0.665 (Download)][IAT c2s/s2c min/avg/max/stddev: 0/0 10/4 200/9197 16/42][Pkt Len c2s/s2c min/avg/max/stddev: 62/62 659/958 1098/1098 393/277][PLAIN TEXT (71pnlF)][Plen Bins: 1,22,20,40,1,1,0,0,0,0,0,0,0,0,0,0,12,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
        3       UDP 10.70.99.11:63916 <-> 147.92.169.110:27647 [proto: 316/LineCall][IP: 315/Line][Encrypted][Confidence: DPI][cat: VoIP/10][12325 pkts/11479132 bytes <-> 16145 pkts/12435762 bytes][Goodput ratio: 95/95][80.13 sec][bytes ratio: -0.040 (Mixed)][IAT c2s/s2c min/avg/max/stddev: 0/0 3/4 283/3695 8/32][Pkt Len c2s/s2c min/avg/max/stddev: 58/62 931/770 1098/1098 299/372][PLAIN TEXT (15yddo)][Plen Bins: 1,12,7,17,2,1,0,0,0,0,0,0,0,0,0,0,13,0,0,0,0,0,0,0,0,0,0,2,1,9,0,21,10,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
        10      UDP 10.70.99.11:54442 <-> 147.92.169.121:11389 [proto: 316/LineCall][IP: 315/Line][Encrypted][Confidence: DPI][cat: VoIP/10][21 pkts/10882 bytes <-> 21 pkts/10779 bytes][Goodput ratio: 92/92][18.97 sec][bytes ratio: 0.005 (Mixed)][IAT c2s/s2c min/avg/max/stddev: 0/10040 44/591 753/10040 177/2362][Pkt Len c2s/s2c min/avg/max/stddev: 73/176 518/513 903/584 187/138][PLAIN TEXT (ihPfnK)][Plen Bins: 2,0,0,0,9,2,2,0,0,0,0,2,2,0,0,0,78,0,0,0,0,0,0,0,0,0,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
        31      UDP 10.70.99.11:53956 <-> 147.92.169.111:10583 [proto: 316/LineCall][IP: 315/Line][Encrypted][Confidence: DPI][cat: VoIP/10][23 pkts/4030 bytes <-> 23 pkts/4679 bytes][Goodput ratio: 76/79][17.60 sec][bytes ratio: -0.075 (Mixed)][IAT c2s/s2c min/avg/max/stddev: 0/0 809/698 2062/2053 824/790][Pkt Len c2s/s2c min/avg/max/stddev: 174/201 175/203 202/257 6/11][Plen Bins: 0,0,0,0,95,2,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
        56      UDP 10.70.99.11:63922 <-> 147.92.169.110:27647 [proto: 316/LineCall][IP: 315/Line][Encrypted][Confidence: DPI][cat: VoIP/10][5 pkts/903 bytes <-> 5 pkts/1066 bytes][Goodput ratio: 77/80][8.10 sec][bytes ratio: -0.083 (Mixed)][IAT c2s/s2c min/avg/max/stddev: 1999/1996 2014/2014 2054/2057 23/25][Pkt Len c2s/s2c min/avg/max/stddev: 175/202 181/213 203/258 11/22][Plen Bins: 0,0,0,0,40,50,10,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]

root@kickseed:~/sharonnDPI/nDPI# ./example/ndpiReader -i /root/linepcap/line_withlaptopinlansetup.pcap  -v2 | grep LineCall
        
LineCall             packets: 98743         bytes: 87328873      flows: 2

        1       UDP 10.70.99.12:56946 <-> 147.92.169.100:17077 [proto: 316/LineCall][IP: 315/Line][Encrypted][Confidence: DPI][cat: VoIP/10][47819 pkts/43400193 bytes <-> 50866 pkts/43917779 bytes][Goodput ratio: 95/95][221.19 sec][bytes ratio: -0.006 (Mixed)][IAT c2s/s2c min/avg/max/stddev: 0/0 3/3 277/6425 7/32][Pkt Len c2s/s2c min/avg/max/stddev: 62/62 908/863 1098/1098 287/327][PLAIN TEXT (dsslcd)][Plen Bins: 1,14,10,0,1,1,1,0,0,0,0,0,0,0,0,0,14,0,0,0,0,0,0,0,0,0,0,2,0,0,0,53,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
        7       UDP 10.70.99.12:56955 <-> 147.92.169.100:17077 [proto: 316/LineCall][IP: 315/Line][Encrypted][Confidence: DPI][cat: VoIP/10][29 pkts/5045 bytes <-> 29 pkts/5856 bytes][Goodput ratio: 76/79][17.58 sec][bytes ratio: -0.074 (Mixed)][IAT c2s/s2c min/avg/max/stddev: 0/0 567/567 2042/2042 730/730][Pkt Len c2s/s2c min/avg/max/stddev: 173/200 174/202 201/256 5/10][Plen Bins: 0,0,0,0,98,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]

```

**Comment by sharonenoch** [comment]
@IvanNardi  Shall I close this? 

I would like to open another enhancement request for 3CX chat and audio/video calls. 
I have the pcaps for the same. Would you be able to take a look and help.?

**Comment by IvanNardi** [comment]
Done in 503aac70bccfecfc920e30b5d7d32e252f0cae92 and 1796a1d814494fa85f75d395703edbc86ead3520

Feel free to open another issue for other protocols, attaching the traces
